File: StereoFlouroscopyRegistration/drr/Cpp/DRR.cxx
#include "itkImage.h"
#include "itkImageFileReader.h"
#include "itkImageFileWriter.h"
#include "itkResampleImageFilter.h"
#include "itkCenteredEuler3DTransform.h"
#include "itkNearestNeighborInterpolateImageFunction.h"
#include "itkImageRegionIteratorWithIndex.h"
#include "itkRescaleIntensityImageFilter.h"
#include "itkFlipImageFilter.h"
#include "itkMatrix.h"

// Software Guide : BeginLatex
//
// The \code{RayCastInterpolateImageFunction} class definition for
// this example is contained in the following header file.
//
// Software Guide : EndLatex

// Software Guide : BeginCodeSnippet
#include "itkRayCastInterpolateImageFunction.h"
// Software Guide : EndCodeSnippet

//#define WRITE_CUBE_IMAGE_TO_FILE


void usage()
{
    std::cerr << "\n";
    std::cerr << "Usage: DRR <options> [input]\n";
    std::cerr << "  calculates the Digitally Reconstructed Radiograph from a volume. \n\n";
    std::cerr << " where <options> is one or more of the following:\n\n";
    std::cerr << "  <-h>                    Display (this) usage information\n";
    std::cerr << "  <-v>                    Verbose output [default: no]\n";
    std::cerr << "  <-res float float>      Pixel spacing of the output image [default: 1x1mm]  \n";
    std::cerr << "  <-size int int>         Dimension of the output image [default: 501x501]  \n";
    std::cerr << "  <-sid float>            Distance of ray source (focal point) [default: 400mm]\n";
    std::cerr << "  <-t float float float>  Translation parameter of the camera \n";
    std::cerr << "  <-rx float>             Rotation around x,y,z axis in degrees \n";
    std::cerr << "  <-ry float>\n";
    std::cerr << "  <-rz float>\n";
    std::cerr << "  <-normal float float>   The 2D projection normal position [default: 0x0mm]\n";
    std::cerr << "  <-cor float float float> The centre of rotation relative to centre of volume\n";
    std::cerr << "  <-threshold float>      Threshold [default: 0]\n";
    std::cerr << "  <-o file>               Output image filename\n\n";
    std::cerr << "                          by  thomas@hartkens.de\n";
    std::cerr << "                          and john.hipwell@kcl.ac.uk (CISG London)\n\n";
    exit(1);
}

int main( int argc, char *argv[] )
{
    char *input_name = ITK_NULLPTR;
    char *output_name = ITK_NULLPTR;
    std::cout << argv[0] << std::endl;
    bool ok;
    bool verbose = false;
    
    float rx = 0.;
    float ry = 0.;
    float rz = 0.;
    
    float tx = 0.;
    float ty = 0.;
    float tz = 0.;
    
    float cx = 0.;
    float cy = 0.;
    float cz = 0.;
    
    float sid = 400.;
    
    float sx = 1.;
    float sy = 1.;
    
    int dx = 501;
    int dy = 501;
    
    float o2Dx = 0;
    float o2Dy = 0;
    
    double threshold=0;
    
    
    
    // Parse command line parameters
    
    while (argc > 1)
    {
        ok = false;
        
        if ((ok == false) && (strcmp(argv[1], "-h") == 0))
        {
            argc--; argv++;
            ok = true;
            usage();
        }
        
        if ((ok == false) && (strcmp(argv[1], "-v") == 0))
        {
            argc--; argv++;
            ok = true;
            verbose = true;
        }
        
        if ((ok == false) && (strcmp(argv[1], "-rx") == 0))
        {
            argc--; argv++;
            ok = true;
            rx=atof(argv[1]);
            argc--; argv++;
        }
        
        if ((ok == false) && (strcmp(argv[1], "-ry") == 0))
        {
            argc--; argv++;
            ok = true;
            ry=atof(argv[1]);
            argc--; argv++;
        }
        
        if ((ok == false) && (strcmp(argv[1], "-rz") == 0))
        {
            argc--; argv++;
            ok = true;
            rz=atof(argv[1]);
            argc--; argv++;
        }
        
        if ((ok == false) && (strcmp(argv[1], "-threshold") == 0))
        {
            argc--; argv++;
            ok = true;
            threshold=atof(argv[1]);
            argc--; argv++;
        }
        
        if ((ok == false) && (strcmp(argv[1], "-t") == 0))
        {
            argc--; argv++;
            ok = true;
            tx=atof(argv[1]);
            argc--; argv++;
            ty=atof(argv[1]);
            argc--; argv++;
            tz=atof(argv[1]);
            argc--; argv++;
        }
        
        if ((ok == false) && (strcmp(argv[1], "-cor") == 0))
        {
            argc--; argv++;
            ok = true;
            cx=atof(argv[1]);
            argc--; argv++;
            cy=atof(argv[1]);
            argc--; argv++;
            cz=atof(argv[1]);
            argc--; argv++;
        }
        
        if ((ok == false) && (strcmp(argv[1], "-res") == 0))
        {
            argc--; argv++;
            ok = true;
            sx=atof(argv[1]);
            argc--; argv++;
            sy=atof(argv[1]);
            argc--; argv++;
        }
        
        if ((ok == false) && (strcmp(argv[1], "-size") == 0))
        {
            argc--; argv++;
            ok = true;
            dx=atoi(argv[1]);
            argc--; argv++;
            dy=atoi(argv[1]);
            argc--; argv++;
        }
        
        if ((ok == false) && (strcmp(argv[1], "-sid") == 0))
        {
            argc--; argv++;
            ok = true;
            sid=atof(argv[1]);
            argc--; argv++;
        }
        
        if ((ok == false) && (strcmp(argv[1], "-normal") == 0))
        {
            argc--; argv++;
            ok = true;
            o2Dx=atof(argv[1]);
            argc--; argv++;
            o2Dy=atof(argv[1]);
            argc--; argv++;
        }
        
        if ((ok == false) && (strcmp(argv[1], "-o") == 0))
        {
            argc--; argv++;
            ok = true;
            output_name = argv[1];
            argc--; argv++;
        }
        
        if (ok == false)
        {
            
            if (input_name == ITK_NULLPTR)
            {
                input_name = argv[1];
                argc--;
                argv++;
            }
            
            else
            {
                std::cerr << "ERROR: Can not parse argument " << argv[1] << std::endl;
                usage();
            }
        }
    }
    
    const     unsigned int   Dimension = 3;
    typedef short InputPixelType;
    typedef short OutputPixelType;
    
    typedef itk::Image<InputPixelType,Dimension> InputImageType;
    InputImageType::Pointer image = InputImageType::New();
    
    typedef itk::Image<OutputPixelType,Dimension> OutputImageType;
    
    typedef itk::ImageFileReader<InputImageType>  ReaderType;
    ReaderType::Pointer reader = ReaderType::New();
    
    reader->SetFileName(input_name);
    
    
    try {
        reader->Update();
    } catch (itk::ExceptionObject & err) {
        std::cerr << "ERROR: ExceptionObject caught !" << std::endl;
        std::cerr << err << std::endl;
        return EXIT_FAILURE;
    }
    image = reader->GetOutput();

    InputImageType::PointType       imOrigin    = image->GetOrigin();
    InputImageType::SizeType        imSize      = image->GetBufferedRegion().GetSize();
    InputImageType::RegionType      imRegion    = image->GetBufferedRegion();
    InputImageType::DirectionType   imDirection = image->GetDirection();
    InputImageType::SpacingType     imSpacing   = image->GetSpacing();
    
    std::cout << image << std::endl;
    
    typedef itk::CenteredEuler3DTransform<double> TransformType;
    TransformType::Pointer transform = TransformType::New();
    
    transform->SetComputeZYX(true);
    TransformType::OutputVectorType translation;
    translation[0] = tx;
    translation[1] = ty;
    translation[2] = tz;
        // constant for converting degrees into radians
    const double dtr = ( std::atan(1.0) * 4.0 ) / 180.0;
    transform->SetTranslation( translation );
    transform->SetRotation( dtr*rx, dtr*ry, dtr*rz );
    
    
    typedef itk::RayCastInterpolateImageFunction<InputImageType,double> InterpolatorType;
    InterpolatorType::Pointer interpolator = InterpolatorType::New();
    
    InterpolatorType::InputPointType focalpoint;
    InputImageType::IndexType desPixelIndx;
    desPixelIndx[0] = (imSize[0]-1.)/2.0; // Find the centre of the volume
    desPixelIndx[1] = (imSize[1]-1.)/2.0; // Find the centre of the volume
    desPixelIndx[2] = (imSize[2]-1.)/2.0; // Find the centre of the volume
    
    image->TransformIndexToPhysicalPoint(desPixelIndx,focalpoint);
    focalpoint[0] = o2Dx;// Location of the principle point in x direction relative to the center of image.
    focalpoint[1] = o2Dy;// Location of the principle point in y direction relative to the center of image.
    focalpoint[2] = sid; // The principal distance or source to image distance.
    
    interpolator->SetTransform(transform);
    interpolator->SetThreshold(threshold);
    interpolator->SetFocalPoint(focalpoint);
    interpolator->Print(std::cout);
    
    
    InputImageType::SizeType size = {{512,512,1}};
    
    typedef itk::ResampleImageFilter<InputImageType, OutputImageType> FilterType;
    FilterType::Pointer filter = FilterType::New();
    
        //----------------------------------------------------------------------------
        // Image Plane Definition : Origin
        //----------------------------------------------------------------------------
    float iOx = 0.0; // The image plane origin in x.
    float iOy = 0.0; // The image plane origin in y.
    float iOz = 0.0; // The image plane origin in z.
    
    desPixelIndx[0] = (size[0]-1.)/2.0;
    desPixelIndx[1] = (size[1]-1.)/2.0;
    desPixelIndx[2] = (size[2]-1.)/1.0;
    
    InputImageType::PointType           outOrigin;
    
    image->TransformIndexToPhysicalPoint(desPixelIndx, outOrigin);
    
    outOrigin[2] = iOz;
//    outOrigin[0] = iOx;
//    outOrigin[1] = iOy;
//    outOrigin[2] = iOz;
//    
    
    
    outOrigin = imDirection*outOrigin; // Projecting it in the desired direction based on the image direction.
    
    InputImageType::SpacingType         outSpacing;
    outSpacing[0] = imSpacing[0];
    outSpacing[1] = imSpacing[1];
    outSpacing[2] = 1.0;
    
    InputImageType::DirectionType      outDirection;
    outDirection = imDirection;
    
    
    filter->SetSize(size);
    filter->SetInterpolator(interpolator);
    filter->SetDefaultPixelValue(0.0);
    filter->SetOutputSpacing(outSpacing);
    filter->SetOutputOrigin(outOrigin);
    filter->SetOutputDirection(imDirection);
    filter->SetTransform(transform);
    filter->SetInput(image);
    
    typedef itk::RescaleIntensityImageFilter<InputImageType,OutputImageType> RescaleFilterType;
    RescaleFilterType::Pointer rescaler = RescaleFilterType::New();
    
    rescaler->SetOutputMinimum(0);
    rescaler->SetOutputMaximum(255);
    rescaler->SetInput(filter->GetOutput());
    
    
    typedef itk::FlipImageFilter< OutputImageType > FlipFilterType;
    FlipFilterType::Pointer flipFilter = FlipFilterType::New();
    
    typedef FlipFilterType::FlipAxesArrayType FlipAxesArrayType;
    FlipAxesArrayType flipArray;
    flipArray[0] = 1; // since X direction was -1 0 0
    flipArray[1] = 1; // since Y direction was 0 -1 0
    
    flipFilter->SetFlipAxes( flipArray );
    flipFilter->SetInput( rescaler->GetOutput() );
    flipFilter->Update();
    
    
    
    typedef itk::ImageFileWriter<OutputImageType> WriterType;
    WriterType::Pointer writer = WriterType::New();
    
    writer->SetFileName(output_name);
    writer->SetInput(flipFilter->GetOutput());
    
    try
    {
    std::cout << "Writing image: " << output_name << std::endl;
    writer->Update();
    }
    catch( itk::ExceptionObject & err )
    {
    std::cerr << "ERROR: ExceptionObject caught !" << std::endl;
    std::cerr << err << std::endl;
    }
    
    
        return EXIT_SUCCESS;
}

File: StereoFlouroscopyRegistration/drr/DRR.py
# This script generates Digitally Reconstructed Radiograph (DRR) from a given 
# 3D image to given 2D image plane(s)

#%% -------------------------------- DEFAULT INPUTS ---------------------------
input_filename  = ''
outputDirectory = ''
outputExtension = '.nii' # Store the output image in nifti format. 
threshold  = 0.
defaultPixelValue = 0
minOut = 0
maxOut = 255

rot = [0.,0.,0.]   # Rotation in degrees in x, y, and z direction. 
trs = [0. ,0. ,0.]   # translation in x, y, and z directions. 
cor = [0. ,0. ,0.]   # offset of the rotation from the center of image (3D)

outSize = [512,512,1] # The size of output image


verbose = False      # Verbose details of all steps. 

nCam = 2 # Number of cameras 




#%% -------------------------------- INPUTS ----------------------------------- 
input_filename  = '/Volumes/Storage/Projects/Registration/QuickData/OA-BEADS-CT.nii'
outputDirectory = '/Volumes/Storage/Projects/Registration/QuickData/'
outputExtension = '.nii' # Store the output image in nifti format. 
threshold  = 0.
defaultPixelValue = 0
minOut = 0
maxOut = 255

rot = [90.,0.,0.]   # Rotation in degrees in x, y, and z direction. 
trs = [10. ,10. ,10.]   # translation in x, y, and z directions. 
cor = [2. ,4. ,6.]   # offset of the rotation from the center of image (3D)

outSize = [512,512,1] # The size of output image


verbose = False      # Verbose details of all steps. 

nCam = 2 # Number of cameras 
calibrationFile = ['/Volumes/Storage/Projects/Registration/QuickData/OAKneeCadaverCd001_NM_Cam1.txt' ,    # Calibration file for Camera 1
                   '/Volumes/Storage/Projects/Registration/QuickData/OAKneeCadaverCd001_NM_Cam2.txt' ]    # Calibration file for Camera 2


#%% ------------------------------- Import Libraries ------------------------
import itk # imports insight Toolkit
import numpy
import sys
import datetime
import Functions
import CalibrationUsingJointTrack 

#%%


if len(calibrationFile) != nCam :
    raise Exception('Number of Calibration files', len(calibrationFile),'do not correspond with the number of Cameras',nCam)

#%%------------------ Starting the main body of the code ---------------- 
# -------------------- Reader -------------------------
InputPixelType  = itk.ctype("short")
OutputPixelType = itk.ctype("short")
ScalarType = itk.D
DimensionIn  = 3
DimensionOut = 3

InputImageType  = itk.Image[InputPixelType , DimensionIn ]
OutputImageType = itk.Image[OutputPixelType, DimensionOut]


ReaderType = itk.ImageFileReader[InputImageType]
reader     = ReaderType.New()
reader.SetFileName(input_filename)

try:
    print("Reading image: " + input_filename)
    reader.Update()
    print("Image Read Successfully")
except ValueError: 
    print("ERROR: ExceptionObject cauth! \n")
    print(ValueError)
    sys.exit()
    
inputImage = reader.GetOutput()

if verbose :
    print(inputImage)
    

#%% ------------------ Transformation 
# This part is inevitable since the interpolator (Ray-cast) and resample Image
# image filter uses a Transformation -- Here we set it to identity. 
TransformType = itk.CenteredEuler3DTransform[itk.D]
transform     = TransformType.New()

transform.SetRotation(numpy.deg2rad(rot[0]),numpy.deg2rad(rot[1]),numpy.deg2rad(rot[2])) # Setting the rotation of the transform
transform.SetTranslation(itk.Vector.D3(trs))    # Setting the translation of the transform
transform.SetComputeZYX(True)  # The order of rotation will be ZYX. 

imOrigin = inputImage.GetOrigin()                   # Get the origin of the image.
inRes    = inputImage.GetSpacing()                  # Get the resolution of the input image.
inSiz    = inputImage.GetBufferedRegion().GetSize() # Get the size of the input image.

center = itk.Point.D3(imOrigin) + numpy.multiply(inRes,inSiz)/2. # Setting the center of rotation as center of 3D object + offset determined by cor. 

transform.SetCenter(center)                     # Setting the center of rotation. 

if verbose :
    print(transform)
    
   #%% 
for ii in range(nCam):
    
    #%%
    imageCalibrationInfo = CalibrationUsingJointTrack.CalibrationTool() # Setting up the image calibration info class. 
    imageCalibrationInfo.SetCalibrationInfo(calibrationFile[ii]) # Assign the information from the calibration file to the imageCalibrationInfo class. 
    
    spaceOutput= imageCalibrationInfo.GetPixelSize() # The resolution (spacing) along x,y,z directions of output image
    
    imageCalibrationInfo.SetOutputImageSize(outSize[0],outSize[1],1)  # Setting the size of the output image. 
    imageCalibrationInfo.SetGlobalOriginForImagePlane()               # Setting the global origin of the output image. 
    
    originOutput = imageCalibrationInfo.GetGlobalOriginForImagePlane() # Setting the output origin. 
    
    directionOutput = imageCalibrationInfo.GetDirectionMatrix() # Direction of Image plane 3x3 matrix. 
    focalPoint = imageCalibrationInfo.GetFocalPoint()  # Position of the x-ray source. 
    
    #%% ----------------- Ray Cast Interpolator 
    # In this part the Ray Cast interpolator is defined and applied to the input
    # image data. 
    
    InterpolatorType = itk.RayCastInterpolateImageFunction[InputImageType,ScalarType]       # Defining the interpolator type from the template. 
    interpolator     = InterpolatorType.New()               # Pointer to the interpolator
    
    interpolator.SetInputImage(inputImage)                  # Setting the input image data
    interpolator.SetThreshold(threshold)                    # Setting the output threshold
    interpolator.SetFocalPoint(itk.Point.D3(focalPoint))    # Setting the focal point (x-ray source location)
    interpolator.SetTransform(transform)                    # Setting the transform (here identity)
    
    if verbose:
        print(interpolator)
    #%%----------------- Resample Image Filter ------------------------
        # In this part the resample image filter to map a 3D image to 2D image plane with desired specs is designed
        
    FilterType = itk.ResampleImageFilter[InputImageType,OutputImageType]                    # Defining the resample image filter type. 
    resamplefilter = FilterType.New()                                                       # Pointer to the filter
    resamplefilter.SetInput(inputImage)                                                     # Setting the input image data 
    resamplefilter.SetDefaultPixelValue(defaultPixelValue)                                  # Setting the default Pixel value
    resamplefilter.SetInterpolator(interpolator)                                            # Setting the interpolator
    resamplefilter.SetTransform(transform)                                                  # Setting the transform
    resamplefilter.SetSize(outSize)                                                         # Setting the size of the output image. 
    resamplefilter.SetOutputSpacing(itk.Vector.D3([spaceOutput[0],spaceOutput[1],1]))       # Setting the spacing(resolution) of the output image. 
    resamplefilter.SetOutputOrigin(originOutput)                                            # Setting the output origin of the image
    Functions.ChangeImageDirection(oldDirection=resamplefilter.GetOutputDirection(),newDirection=directionOutput,DimensionOut=3)     # Setting the output direction of the image  --- resamplefilter.SetImageDirection(args) was not working properly
    
    resamplefilter.Update()                                                                 # Updating the resample image filter.
    
    if verbose:
        print(resamplefilter)
    #%%---------------- Rescaler Image Filter --------------------------
    RescalerFilterType = itk.RescaleIntensityImageFilter[InputImageType,OutputImageType]    # Defining the rescale image filter. 
    rescaler = RescalerFilterType.New()             # Pointer to the rescale filter
    rescaler.SetOutputMinimum(minOut)               # Minimum output
    rescaler.SetOutputMaximum(maxOut)               # Maximum output 
    rescaler.SetInput(resamplefilter.GetOutput())   # Setting the input to the image filter. 
    rescaler.Update() 
    
    if verbose:
        print(rescaler)
    
    #%% ------------------ Writer ------------------------------------
        # The output of the resample filter can then be passed to a writer to
        # save the DRR image to a file.
    WriterType = itk.ImageFileWriter[OutputImageType]
    writer = WriterType.New()
    
    outputPath = '/Volumes/Storage/Projects/Registration/QuickData/Cam'+str(ii+1)
    if ii == 0:
        time = datetime.datetime.now() 
        
    outputName = ('/Cam'+str(ii+1)+'rx'+str(int(rot[0]))+'ry'+str(int(rot[1]))+'rz'+str(int(rot[2]))+'tx'
                 + str(int(trs[0]))+'ty'+str(int(trs[1]))+'tz'+str(int(trs[2]))+'y'+str(time.year)+'m'+str(time.month)
                 +'d'+str(time.day)+'hr'+str(time.hour)+'m'+str(time.minute)+'s'+str(time.second)+ outputExtension)
    output_filename = outputPath+outputName
    
    writer.SetFileName(output_filename)
    writer.SetInput(rescaler.GetOutput())
    
    try:
        print("Writing image: " + output_filename)
        writer.Update()
        print("Image Printed Successfully")
    except ValueError: 
        print("ERROR: ExceptionObject cauth! \n")
        print(ValueError)
        sys.exit()
        